common/json-parser.c

```diff
@@ -0,0 +1,1290 @@
+ *  union {string, d_number, i_number, boolean, comp}:
+ *      string,d_number,i_number,boolean: item literal value, denoted 
+ *      by its type.  */
+typedef struct json_item_s {
+    union {
+        char *string;
+        long double number;
+        _Bool boolean;
+        json_composite_t *comp;
+    };
+    enum json_type type;
+
+    char *key;
+    struct json_item_s *parent;
+} json_item_t;
+
+struct _parse_context {
+    char *buffer;
+    char *key; /* holds key ptr to be received by item */
+};
+
+
+/* function pointers used while building json items, 
+    json_create_value points to functions prefixed by "_json_value_set_"
+    json_create_item points to functions prefixed by "json_decode" */
+typedef void (json_create_value)(json_item_t *item, struct _parse_context *cxt);
+typedef json_item_t* (json_create_item)(json_item_t*, struct _parse_context*, json_create_value*);
+
+static json_item_t*
+_json_item_init()
+{
+    json_item_t *new_item = calloc(1, sizeof *new_item);
+    ASSERT_S(NULL != new_item, "Out of memory");
+
+    return new_item;
+}
+
+/* create a new branch to current json object item, and return
+    the new branch address */
+static json_item_t*
+_json_branch_init(json_item_t *item)
+{
+    ++item->comp->num_branch;
+
+    item->comp->branch[item->comp->num_branch-1] = _json_item_init();
+
+    item->comp->branch[item->comp->num_branch-1]->parent = item;
+
+    return item->comp->branch[item->comp->num_branch-1];
+}
+
+static void
+_json_composite_destroy(json_item_t *item)
+{
+    free(item->comp->branch);
+    item->comp->branch = NULL;
+
+    free(item->comp);
+    item->comp = NULL;
+}
+
+static void
+_json_destroy_preorder(json_item_t *item)
+{
+    switch (item->type){
+    case JSON_OBJECT:
+    case JSON_ARRAY:
+        for (size_t i=0; i < item->comp->num_branch; ++i){
+            _json_destroy_preorder(item->comp->branch[i]);
+        }
+        _json_composite_destroy(item);
+        break;
+    case JSON_STRING:
+        free(item->string);
+        item->string = NULL;
+        break;
+    default:
+        break;
+    }
+
+    if (NULL != item->key){
+        free(item->key);
+        item->key = NULL;
+    }
+
+    free(item);
+    item = NULL;
+}
+
+/* destroy current item and all of its nested object/arrays */
+void
+json_destroy(json_item_t *item){
+    _json_destroy_preorder(json_get_root(item));
+}
+
+static json_composite_t*
+_json_decode_composite(char **p_buffer, size_t n_branch){
+    json_composite_t *new_comp = calloc(1, sizeof *new_comp);
+    ASSERT_S(NULL != new_comp, "Out of memory");
+
+    new_comp->branch = malloc((1+n_branch) * sizeof(json_item_t*));
+    ASSERT_S(NULL != new_comp->branch, "Out of memory");
+
+    ++*p_buffer; /* skips composite's '{' or '[' delim */
+
+    return new_comp;
+}
+
+static char*
+_json_decode_string(char **p_buffer)
+{
+    char *start = *p_buffer;
+    ASSERT_S('\"' == *start, "Not a string");
+
```

Comment:
Can you please file an issue to replace this with jsmn's parser later?  It's better we only keep one parser in the code base to reduce the complexity and maintenance burden.

Reply:
Completely agree, I will file an issue.

Reply:
It will also help reduce a lot of code.